controllers/horizontal_runner_autoscaler_webhook.go

```diff
@@ -716,31 +795,55 @@ func (autoscaler *HorizontalRunnerAutoscalerGitHubWebhook) SetupWithManager(mgr
 		switch hra.Spec.ScaleTargetRef.Kind {
 		case "", "RunnerDeployment":
 			var rd v1alpha1.RunnerDeployment
-
 			if err := autoscaler.Client.Get(context.Background(), types.NamespacedName{Namespace: hra.Namespace, Name: hra.Spec.ScaleTargetRef.Name}, &rd); err != nil {
+				autoscaler.Log.V(1).Info(fmt.Sprintf("RunnerDeployment not found with scale target ref name %s for hra %s", hra.Spec.ScaleTargetRef.Name, hra.Name))
 				return nil
 			}
 
-			keys := []string{rd.Spec.Template.Spec.Repository, rd.Spec.Template.Spec.Organization}
-
+			keys := []string{}
+			if rd.Spec.Template.Spec.Repository != "" {
+				keys = append(keys, rd.Spec.Template.Spec.Repository) // Repository runners
+			}
+			if rd.Spec.Template.Spec.Organization != "" {
+				if group := rd.Spec.Template.Spec.Group; group != "" {
+					keys = append(keys, organizationalRunnerGroupKey(rd.Spec.Template.Spec.Organization, rd.Spec.Template.Spec.Group)) // Organization runner groups
+				} else {
+					keys = append(keys, rd.Spec.Template.Spec.Organization) // Organization runners
+				}
+			}
 			if enterprise := rd.Spec.Template.Spec.Enterprise; enterprise != "" {
-				keys = append(keys, enterpriseKey(enterprise))
+				if group := rd.Spec.Template.Spec.Group; group != "" {
+					keys = append(keys, enterpriseRunnerGroupKey(enterprise, rd.Spec.Template.Spec.Group)) // Enterprise runner groups
```

Comment:
It turns out we don't need to find HRA by runner groups as GitHub finds the runner to scheduled a job by labels anyway.
What we need instead would be to list all the HRAs whose RunnerDeployment matches the workflow_job labels and filter out those that are associated to runner groups (both organization and enterprise groups) that are unavailable to the org / the repo.

Reply:
@mumoshu but you can use groups without labels, you coukd set runner groups with custom visibility and no custom labels right? (That's a common pattern in our company) if we dont take group into consideration, we could scale up the wrong runner group

Reply:
That's not really the design philosophy behind runner groups though. If Github had intended on them as a means for defining exclusive access they would have:

1. included the repository runner group membership information in the workflow_jobs event whilst queued
2. enabled you to remove repositories from being members of the `Default` runner group.

Runner groups are a means for _extending_ what runners are avaliable to a repository via provided access via labels. They aren't designed to be a way of defining placement of runners beyond opening up access different classes of runners via labels which in turn implicitly defines placement.

Reply:
> 1. included the repository runner group membership information in the workflow_jobs event whilst queued

This is one of the points that we're dicussing with GitHub (still pending) to potentially include that information in the webhook payload and have the full support as they do support custom visibility. That could simplify some of the logic here to avoid querying GitHub API and find the matching runner group

> 2. enabled you to remove repositories from being members of the `Default` runner group.

You can not add runners to the default runner group (this is what we're doing) that way qe preserve the intent of having only repository specific runner groups

> Runner groups are a means for _extending_ what runners are avaliable to a repository via labels. They aren't designed to be a way of defining placement of runners beyond opening up access different classes of runners via labels which in turn implicitly defines placement.

Uhm, that's why GitHub added the custom visibility isn't it? So that you could define some sort of protected runners yo a specific set of repositories. Labels won't do anything w.r.t protection or scoping some runners to some repositories only.

Reply:
> @mumoshu but you can use groups without labels, you could set runner groups with custom visibility and no custom labels right? (That's a common pattern in our company) if we dont take group into consideration, we could scale up the wrong runner group

@fgalind1 Correct. But this seems like considering it in a wrong way. What's implemented in this PR only works when you only use a lot of runner groups with the same set of labels.

If you configure a different set of labels per runner groups, what's done in this PR scales wrong runner runnerdeployments.

Reply:
Instead you'd need to do what I've explained in https://github.com/actions-runner-controller/actions-runner-controller/pull/1012#discussion_r785624849

Reply:
> I believe what we had to do instead was to skip this target(=a runnerdepoyment that is associated to an organization) if it was associated to a runner group and that runner group was indivisible to this repository.

@mumoshu just to clarify with your previous comment, right now we find the potential HRA with the following precedence:

1. matching runner in repo from webhook
2. matching default runner in an org from webhook
3. matching default runner in enterprise from webhook
4. matching runner group in an org that matches repo from webhook and custom visibility of runner group
5. matching runner group in an org that matches repo from webhook and custom visibility of runner group

with your comment - are you suggesting to remove/skip 2 and 3? 

> it was associated to a runner group 

This is the part that wasn't clear to me what object where you referring to be associated to the runner group

Reply:
I'd say it should be like the below one instead:

1. matching repository-wide runner from the repository, the default org runner group, and the non-default org runner group that are visible to the repository, from webhook
2. matching organizational runner from the organization, the default enterprise runner group, and the non-default enterprise runner groups that are visible to the organization, from webhook
3. matching default runner in enterprise from webhook

You may already know but there're two variants of runner groups, organizational runner groups and enterprise runner groups. The visibility for the former can be controlled per repository, not organization. Similarly, the visibility for the latter can be controlled per organization under the enterprise.